fix/gifted-pets.lua

```diff
@@ -0,0 +1,112 @@
+-- Fixes pets gifted to another unit in Adventure Mode
+-- Fixes pets owned by companions missing from the companions list
+
+local function isPetOf(petNemesis, ownerNemesis)
+    if petNemesis.unit then
+        local owner_id = petNemesis.unit.relationship_ids.Pet
+        if owner_id ~= -1 and owner_id == ownerNemesis.unit_id then
+            return true
+        end
+    elseif petNemesis.figure then -- in case the unit is offloaded
+        for _, link in ipairs(petNemesis.figure.histfig_links) do
+            if link._type == df.histfig_hf_link_pet_ownerst and ownerNemesis.figure and link.target_hf == ownerNemesis.figure.id then
+                return true
+            end
+        end
+    end
+    return false
+end
+
+
+local function findPetOwnerOf(petNemesis)
+    if petNemesis.figure then
+        for _, link in ipairs(petNemesis.figure.histfig_links) do
+            if link._type == df.histfig_hf_link_pet_ownerst then
+                local target_hf = link.target_hf
+                return df.nemesis_record.find(df.historical_figure.find(target_hf).nemesis_id)
+            end
+        end
+    end
+    return nil
+end
+
+local function findInParty(party, histfig_id)
+    for k, v in ipairs(party) do
+        if v == histfig_id then
+            return true
+        end
+    end
+    return false
+end
+
+local function fixMissingCompanionPets(nemesis)
+    local party = df.global.adventure.interactions
+    for i, nem_id in ipairs(nemesis.companions) do
+        local pet_record = df.nemesis_record.find(nem_id)
+        if isPetOf(pet_record, nemesis) and not findInParty(party.party_pets, pet_record.figure.id) then
+            print("inserting pet "..pet_record.id.." of companion "..nemesis.id.." into the party!")
+            party.party_pets:insert('#', pet_record.figure.id)
+        end
+    end
+end
+
+
+local function fixGiftedPet(pet_record, pet_owner_record)
+    -- associate the unit relationships
+    if pet_record.unit then
+        pet_record.unit.relationship_ids.Pet = pet_owner_record.unit_id
+        pet_record.unit.relationship_ids.GroupLeader = pet_owner_record.unit_id
+        pet_record.group_leader_id = pet_owner_record.id
+    end
+    
+    -- associate the site links
+    local site_links = pet_owner_record.figure.site_links
+    for u, link in ipairs(site_links) do
+        if link._type == df.histfig_site_link_home_site_abstract_buildingst or
+        link._type == df.histfig_site_link_home_site_realization_buildingst or
+        link._type == df.histfig_site_link_lairst or
+        link._type == df.histfig_site_link_home_site_realization_sulst or
+        link._type == df.histfig_site_link_home_site_saved_civzones then
+            -- Make a copy of the link
+            local new_site_link = df.new(link)
+            -- Assign the same data
+            new_site_link.site = link.site
+            new_site_link.sub_id = link.sub_id
+            new_site_link.entity = link.entity
```

Comment:
should already be done by the `df.new(link)` (maybe make that `link:new()` to make it clearer)